refactor: run the webserver unprivileged

sgaist opened the pull request:
<!--  Thanks for sending a pull request! Here are some tips for you:
1. If this is your first time, please read our contributor guidelines in the https://github.com/falcosecurity/.github/blob/main/CONTRIBUTING.md file.
2. Please label this pull request according to what type of issue you are addressing.
3. Please add a release note!
4. If the PR is unfinished while opening it specify a wip in the title before the actual title, for example, "wip: my awesome feature"
-->

**What type of PR is this?**

> Uncomment one (or more) `/kind <>` lines:

> /kind bug

/kind cleanup

/kind design

> /kind documentation

> /kind failing-test

/kind feature

> /kind release

<!--
Please remove the leading whitespace before the `/kind <>` you uncommented.
-->

**Any specific area of the project related to this PR?**

> Uncomment one (or more) `/area <>` lines:

> /area build

> /area engine

> /area tests

> /area proposals

> /area CI

<!--
Please remove the leading whitespace before the `/area <>` you uncommented.
-->

**What this PR does / why we need it**:

The webserver is running in a thread with the same privileges as the falco process. In order to make running it more secure, use fork rather than thread and then change the user and group id to unprivileged values.

**Which issue(s) this PR fixes**:

<!--
Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
If PR is `kind/failing-tests` or `kind/flaky-test`, please post the related issues/tests in a comment and do not use `Fixes`.
-->

Fixes #3485 

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

<!--
If NO, just write "NONE" in the release-note block below.

If YES, a release note is required, enter your release note in the block below. 
The convention is the same as for commit messages: https://github.com/falcosecurity/.github/blob/main/CONTRIBUTING.md#commit-convention
If the PR introduces non-backward compatible changes, please add a line starting with "BREAKING CHANGE:" and describe what changed.
For example, `BREAKING CHANGE: the API interface of the rule engine has changed`.
Your note will be included in the changelog.
-->

```release-note
webserver: the webserver component now runs in an unprivileged fork. The user id and group id used to run the fork can be configured through the webserver configuration block.
```

sgaist committed:
refactor: run the webserver unprivileged

The webserver was running in a thread with the same
privileges as the falco process. In order to make running
it more secure, use fork rather than thread and then change
the user and group id to unprivileged values.

Signed-off-by: Samuel Gaist <samuel.gaist@idiap.ch>

sgaist commented:
Note for reviewers: it's the first time I am taking a shot at fork so I might have implemented things in a wrong fashion.
I will welcome hints to documentation/best practice for it.

ekoops commented on userspace/falco/webserver.cpp:
```diff
@@ -58,47 +59,62 @@ void falco_webserver::start(const falco::app::state &state,
+		int res = setgid(webserver_config.m_uid);
+		if(res != NOERROR) {
+			throw falco_exception("Webserver: an error occurred while setting group id: " +
+			                      std::to_string(errno));
```
Could we use `std::strerror()` here and in the other occurrences?

ekoops commented on userspace/falco/webserver.cpp:
```diff
@@ -58,47 +59,62 @@ void falco_webserver::start(const falco::app::state &state,
 			falco_logger::log(falco_logger::level::ERR,
-			                  "falco_webserver: " + std::string(e.what()) + "\n");
+			                  "Webserver: " + std::string(e.what()) + "\n");
+			m_failed.store(true, std::memory_order_release);
```
I guess this is not gonna work... The child process updates its own `m_failed` copy. In order to establish a communication between the two processes, we should use an IPC mechanism, like a pipe. In this way, the child process, could communicate to falco that it correctly bootstrapped the webserver or that it failed while attempting.

sgaist replied:
You're correct, I had forgotten about that rule.

I have implemented the check using `cpp-httplib`'s client.

ekoops commented:
Hey @sgaist , any update on this?

leogr commented:
@sgaist any update on this?

sgaist committed:
refactor: use client to check if the web server is running

Using a local variable won't work since the server and the
main application don't share the same memory space anymore.

To check that the server is running, a call to the health
endpoint is used.

While the start of the server should be pretty fast, a
simple exponential backoff has been implemented to allow
for room in case the start is slowed for some reason.

Signed-off-by: Samuel Gaist <samuel.gaist@idiap.ch>

sgaist committed:
fix: use std::strerror rather than just errno

Signed-off-by: Samuel Gaist <samuel.gaist@idiap.ch>

sgaist committed:
fix: do not throw from a nothrow function

Signed-off-by: Samuel Gaist <samuel.gaist@idiap.ch>

ekoops commented on userspace/falco/webserver.cpp:
```diff
@@ -58,47 +61,85 @@ void falco_webserver::start(const falco::app::state &state,
+		int res = setgid(webserver_config.m_uid);
+		if(res != NOERROR) {
+			throw falco_exception(
+			        std::string("Webserver: an error occurred while setting group id: ") +
```
Here we are calling setgid but we are providing m_uid. If the keep setgid here, then the exception message is ok; otherwise we also need to change it

ekoops commented on userspace/falco/webserver.cpp:
```diff
@@ -58,47 +61,85 @@ void falco_webserver::start(const falco::app::state &state,
 		}
-		if(m_server_thread.joinable()) {
-			m_server_thread.join();
+		falco_logger::log(falco_logger::level::INFO, "Webserver: killing fork\n");
```
What about using "webserver process" instead of "fork"? I mean, both here and in the other places... WDYT?

sgaist committed:
fix: use correct parameter for setgid and setuid

They were inverted.

Signed-off-by: Samuel Gaist <samuel.gaist@idiap.ch>

sgaist committed:
refactor: use configured listen address for liveness check

Signed-off-by: Samuel Gaist <samuel.gaist@idiap.ch>

sgaist committed:
refactor: reword information

Improve readability

Signed-off-by: Samuel Gaist <samuel.gaist@idiap.ch>

sgaist commented on userspace/falco/webserver.cpp:
```diff
@@ -58,47 +61,85 @@ void falco_webserver::start(const falco::app::state &state,
+		if(webserver_config.m_ssl_enabled) {
+			schema = "https";
+		}
+		std::string url = schema + "://localhost:" + std::to_string(webserver_config.m_listen_port);
```
Good point

Done

ekoops commented on userspace/falco/webserver.cpp:
```diff
@@ -58,47 +61,88 @@ void falco_webserver::start(const falco::app::state &state,
 		} catch(std::exception &e) {
 			falco_logger::log(falco_logger::level::ERR,
-			                  "falco_webserver: " + std::string(e.what()) + "\n");
+			                  "Webserver: " + std::string(e.what()) + "\n");
```
My concern is what happens if the `m_server->listen(...)` call fail. Here, we are catching the exception, and after we print a log line, we just return from the function. Even if the main falco process will recognize that the son is not running, I guess we should avoid letting the web server process to proceed.

sgaist replied:
Currently, if the server does not start for some reason, it will be detected by the loop below that tries to connect to the server. However, it's not a nice way to detect that.
I have refactored that part.

sgaist committed:
fix: use nobody uid/gid as default values for webserver

Signed-off-by: Samuel Gaist <samuel.gaist@idiap.ch>

sgaist committed:
refactor(webserver): improve failure handling

If starting the server throws for some reason,
the result will be instantly shown.

Signed-off-by: Samuel Gaist <samuel.gaist@idiap.ch>

sgaist commented:
I was able to come back to this and I think it needs a bit more design brainstorming (I was I bit tired when starting this and there are several details that escaped me at that time).

I think that the webserver, as a forked process, should not have anything to do with directly accessing the `state` object to retrieve the stats. We should rather have a proper IPC in place that decouples both.

What do you think ?

poiana commented:
Issues go stale after 90d of inactivity.

Mark the issue as fresh with `/remove-lifecycle stale`.

Stale issues rot after an additional 30d of inactivity and eventually close.

If this issue is safe to close now please do so with `/close`.

Provide feedback via https://github.com/falcosecurity/community.

/lifecycle stale

leogr commented:
/remove-lifecycle stale

/milestone 0.46.0
